lambdajs/src/SharedInterpreter.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE Arrows                     #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TypeSynonymInstances       #-}
+
+{-# LANGUAGE FunctionalDependencies     #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+module SharedInterpreter where
+
+import           Control.Arrow
+import           Control.Arrow.Utils (map)
+import           Prelude             hiding (break, error, lookup, map, read)
+import qualified Prelude
+import           Syntax
+
+class Arrow c => AbstractValue v c | c -> v where
```

Comment:
The name `AbstractValue` is confusing. Someone could think this type class is only for abstract interpreters.

Reply:
Would it make sense to break up this type class into smaller type classes to make them more reusable?